End the deprecation period of using the result of a comma expression.

wilzbach opened the pull request:
The deprecation period could have ended a lot earlier.
DMD deprecations should really be tracked at a common place, s.t. for every
release cycle old deprecations can be bumped accordingly.

dlang.org: https://github.com/dlang/dlang.org/pull/2151

wilzbach committed:
End the deprecation period of using the result of a comma expression.

quickfur commented:
I wonder if there's a way to use `libdparse`, say, to collect all `deprecated` symbols into a list somewhere, so that we can review which symbols need to be bumped to the next stage of deprecation.

Depending on what we find in such a list, we might even be able to automatically parse the deprecation dates and output a list of symbols that need to be updated.  In the distant (or not-so-distant?) future we could even have a script/bot that automatically submits PRs to this effect.  Automation FTW!

quickfur commented on changelog/comma-deprecation-error.dd:
```diff
@@ -0,0 +1,4 @@
+The deprecation period of using the result of comma expression has ended
+
+Comma expressions have proven to be a frequent source of confusion, and bug.
```
Nitpick: `bug` > `bugs`.

quickfur commented on src/dmd/expressionsem.d:
```diff
@@ -5844,7 +5844,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
         e.type = e.e2.type;
         if (e.type !is Type.tvoid && !e.allowCommaExp && !e.isGenerated)
-            e.deprecation("Using the result of a comma expression is deprecated");
+            e.deprecation("The result of a comma expression cant be used");
```
Nitpick: `cant` > `can't`.

quickfur commented on src/dmd/expressionsem.d:
```diff
@@ -6272,7 +6272,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
              */
             if (!(cast(CommaExp)exp.e2).isGenerated)
-                exp.deprecation("Using the result of a comma expression is deprecated");
+                exp.error("The result of a comma expression cant be used");
```
ditto

wilzbach replied:
Well cant is the existing DMD style:

```
> ag "cant" | wc -l
466 
> ag "can't" | wc -l
128
```

quickfur replied:
Existing bad grammar should not be justification for adding more bad grammar!

wilzbach replied:
I have seen @WalterBright repeatedly use "cant" - even in recent PRs. That was my motivation to use "cant".

wilzbach replied:
Examples:
- https://github.com/dlang/dmd/pull/6092
- https://github.com/dlang/dmd/pull/7307

JinShil replied:
"cant" is not a word.  I believe @WalterBright made a mistake in those PRs.

andralex replied:
Yah let's fix those. Thanks!

wilzbach replied:
Fair enough, I changed it to can't.

quickfur replied:
@JinShil Actually, "cant" *is* a word... but it has a [different meaning](https://en.wiktionary.org/wiki/cant). :-D 

@wilzbach The next time Walter submits a PR with the wrong use of "cant", somebody should point it out so that it can be fixed. :-P

quickfur commented on test/fail_compilation/commaexp.d:
```diff
@@ -1,13 +1,13 @@
+fail_compilation/commaexp.d(38): Error: The result of a comma expression cant be used
+fail_compilation/commaexp.d(39): Error: The result of a comma expression cant be used
+fail_compilation/commaexp.d(41): Error: The result of a comma expression cant be used
+fail_compilation/commaexp.d(42): Error: The result of a comma expression cant be used
```
ditto

quickfur commented on test/fail_compilation/diag10862.d:
```diff
@@ -57,7 +57,7 @@ fail_compilation/diag10862.d(77): Error: assignment cannot be used as a conditio
 fail_compilation/diag10862.d-mixin-81(81): Error: assignment cannot be used as a condition, perhaps `==` was meant?
 fail_compilation/diag10862.d-mixin-82(82): Error: assignment cannot be used as a condition, perhaps `==` was meant?
-fail_compilation/diag10862.d-mixin-83(83): Deprecation: Using the result of a comma expression is deprecated
+fail_compilation/diag10862.d-mixin-83(83): Error: The result of a comma expression cant be used
```
ditto

quickfur commented on test/fail_compilation/ice10949.d:
```diff
@@ -1,7 +1,7 @@
 TEST_OUTPUT:
 ---
-fail_compilation/ice10949.d(12): Deprecation: Using the result of a comma expression is deprecated
+fail_compilation/ice10949.d(12): Error: the result of a comma expression cant be used
```
ditto

quickfur commented on test/runnable/ctorpowtests.d:
```diff
@@ -136,7 +136,8 @@ static assert(bazra(14)==64);
 {
-   auto k = (containsAsm(), containsAsm());
+   (containsAsm(), containsAsm());
+   auto k = containsAsm();
```
Huh? Shouldn't this be:
````
containsAsm();
auto k = containsAsm();
````
instead?  This change doesn't make sense to me.

quickfur replied:
Ah, I see why now. Nevermind, carry on. :-P

wilzbach commented:
> I wonder if there's a way to use libdparse, say, to collect all deprecated symbols into a list somewhere, so that we can review which symbols need to be bumped to the next stage of deprecation.

I like the idea, but it would be a lot easier for Phobos. For dmd you are basically left with `grep` for now.
OTOH even for Phobos `grep` works fine - we even have the `@@@DEPRECATED_` pattern for Phobos, which we don't have here.
Hence, my proposal to have a simple file or wiki in which this is being kept track of for dmd - essentially it's the table at https://dlang.org/deprecate.html, but just with more love (not many deprecations actually update this file and sometimes there are compiler bugs).

https://github.com/dlang/dmd/pull/7760/files is one of the first PRs to use the deprecated pattern in dmd too.

wilzbach commented:
So this is finally passing the testsuite :rocket: and just in case someone was wondering obviously my motivation is to make this an error with 2.079, s.t. comma expressions can be re-purposed for built-in tuples.
See also: [upcoming tuple DIP](https://forum.dlang.org/post/p3bdp1$2b4e$1@digitalmars.com)

JinShil commented on test/runnable/testscope.d:
```diff
@@ -294,7 +294,7 @@ void test11()
     bar11(p, &i);
 
-    bar11((i,p), &i);  // comma expressions are deprecated, but need to test them
+    bar11(p, &i);
```
This test was specifically testing the comma operator.  Now that it produces an error, it should probably be moved to the fail_compilation tests.  However, given that the equivalent expression already exists on line 295, it would probably be ok, to just delete it.

wilzbach replied:
Yeah, I was very careful not to delete anything, but yeah in this case it does make sense. Thanks!

JinShil commented on test/runnable/xtest46.d:
```diff
@@ -7866,7 +7867,7 @@ bool test16022()
 {
```
This [issue 16022](https://issues.dlang.org/show_bug.cgi?id=16022) was very specific to the comma operator.  Since the result of a comma expression now emits an error, this test should no longer compile, and should probably be moved to the fail_compilation test suite.

wilzbach replied:
Ok. I copied it to fail compilation. I don't see any reason to remove it though.

JinShil commented on src/dmd/expressionsem.d:
```diff
@@ -6272,7 +6272,7 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
              */
             if (!(cast(CommaExp)exp.e2).isGenerated)
-                exp.deprecation("Using the result of a comma expression is deprecated");
+                exp.error("The result of a comma expression can't be used");
```
I'd prefer this to say "The result of a comma expression is deprecated and can no longer be used".  But, it's ok the way it is; I'll pull it regardless.

wilzbach replied:
Hmm it's not deprecated anymore, it's a real error and should be displayed to the user as such.

jacob-carlborg replied:
I agree with @wilzbach.

CyberShadow replied:
Would be nice to distinguish "can't be used" (due to some technical limitation) from "can't be used" (because we decided to forbid it). How about "Using the result of a comma expression is not allowed" ?

jacob-carlborg commented on changelog/comma-deprecation-error.dd:
```diff
@@ -0,0 +1,4 @@
+The deprecation period of using the result of comma expression has ended
+
+Comma expressions have proven to be a frequent source of confusion, and bugs.
```
Please include an example that now fails to compile and show how to correct it.

jacob-carlborg replied:
Perhaps include an example showing how the comma expression can still be used without using the result.

wilzbach replied:
I was hoping I could avoid the redundancies from https://dlang.org/deprecate.html#Using%20the%20result%20of%20a%20comma%20expression, but okay copying doesn't cost us much.
I copied the text I have written for dlang.org a while ago.

jacob-carlborg commented on changelog/comma-deprecation-error.dd:
```diff
@@ -0,0 +1,59 @@
+    foo((++a, b));
+
+    synchronized (lockA, lockB) {}
+    // isn't currently implemented, but still compiles "thanks" to the comma operator
```
`"thanks"` sounds a bit condescending. Perhaps replace with `due`.

wilzbach replied:
Also submitted for dlang.org: https://github.com/dlang/dlang.org/pull/2152

jacob-carlborg commented on test/runnable/mangle.d:
```diff
@@ -324,23 +324,25 @@ void test9525()
         a(&s);  // Error: Cannot convert &S to const(S*) at compile time
+        return true;
     }
-    static assert((test2!f9525(), true));
```
Are the `static assert` not necessary anymore? Or will the `static assert` inside the function take care of that?

wilzbach replied:
I think the `static assert` was only there, s.t. the function got evaluated and the `static assert` inside of it could trigger, hence I added `evalTest1` and `evalTest2`.

jacob-carlborg replied:
Ok.

CyberShadow commented on changelog/comma-deprecation-error.dd:
```diff
@@ -0,0 +1,59 @@
+    ---
+
+    It's also common to use the comma operator in for-loop increment
+    statements to allow multiple expressions.
```
I think it should be clearer that this use case is not affected.

CyberShadow commented on changelog/comma-deprecation-error.dd:
```diff
@@ -0,0 +1,59 @@
+)
+$(H4 Rationale)
+$(P The comma operator leads to unintended behavior (see below for a selection)
+    Moreover it is not commonly used and it blocks the ability to implement tuples.
```
It blocks implementing tuples as a language feature using commas, not in general

CyberShadow commented on changelog/comma-deprecation-error.dd:
```diff
@@ -0,0 +1,59 @@
+$(P The comma operator leads to unintended behavior (see below for a selection)
+    Moreover it is not commonly used and it blocks the ability to implement tuples.
+
+    A selection of problems through the accidental use of the comma operator:
```
Honestly all of these examples are entirely unconvincing because they could be caught by a "discarding result of side-effect-free subexpression" warning.

wilzbach replied:
It took them from here: https://dlang.org/deprecate.html#Using%20the%20result%20of%20a%20comma%20expression
https://github.com/dlang/dlang.org/pull/1308

If you have better examples, just let me know.

CyberShadow commented on changelog/comma-deprecation-error.dd:
```diff
@@ -0,0 +1,59 @@
+        // will always be reached
+    }
+
+    void foo(int x) {}
```
I don't understand this example. Perhaps it would make sense for `void foo(int x, int y=0)` ?

wilzbach commented:
Addressed all comments except for:

> Honestly all of these examples are entirely unconvincing because they could be caught by a "discarding result of side-effect-free subexpression" warning.

It took them from here: https://dlang.org/deprecate.html#Using%20the%20result%20of%20a%20comma%20expression
dlang/dlang.org#1308
If you have better examples, just let me know. And we can always add them to changelog later.

wilzbach commented:
@JinShil you still have this on "requested changes". I addressed your comments and copied the files over to `fail_compilation`, but kept the modified version in `compilation`. Are you okay with this?
I would like to move on with this, s.t. we can sip this with 2.079 and can introduce tuples into D as soon as possible ...

quickfur commented:
Woohoo!!!! The comma operator is officially dead!!! *celebration*
